### PR TITLE
Feature: Implement MSIZE opcode

### PR DESCRIPTION
**Pull request by spartucus**
Spec reference from [MSIZE](https://github.com/appliedzkp/zkevm-specs/blob/master/specs/opcode/59MSIZE.md).

### CODE REVIEW

File: bus-mapping/src/evm/memory.rs
Comment (CPerezz) [inline bus-mapping/src/evm/memory.rs:276]: Maybe `self.0.len() >> 5` does the truncation.

Also, you can make the fn `const` or if not possible, at least make it `#[inline]`.

**Review by CPerezz — APPROVED** [review]
LGTM.

Let's wait for one more approval! Good job! :smile:

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/msize.rs:106]: Would the simplification applied here https://github.com/appliedzkp/zkevm-circuits/pull/239/files#diff-6cab7856d5161af3600f3ba7b241c0f6bb26ff41ee8d8a04595212243e76710bR93
work here as well?

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/execution/msize.rs:33]: I think `memory_size` is capped at 32 bits due to quadratic part of memory expansion gas cost, but I am also planning to file a PR to make this part consistent with spec, we could update this then since there are other trivial things need to be change at the same time.

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (han0110) [inline zkevm-circuits/src/evm_circuit/execution/msize.rs:80]:
```suggestion
            Some(step.memory_size.to_le_bytes()),
```

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:106]: I think this PR needs rebase to work, some structures have been moved to another module.

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (spartucus) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:33]: I need some advice here, since every `assign_exec_step` is called in `execution.rs` by different execution_state. For memory_size, in current step it has previous step's memory_size, but that's not we want. See code [here](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/witness.rs#L388-L403).

File: bus-mapping/src/evm/memory.rs
Comment (spartucus) [reply bus-mapping/src/evm/memory.rs:276]: >Maybe self.0.len() >> 5 does the truncation.

It is same as '/ 32', we need take care situations that divided by 32 left a reminder.

>Also, you can make the fn const or if not possible, at least make it #[inline].

`#[inline]` sound good.

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:33]: I was not aware of this behavior from geth!  What we could do is modify the geth Execution Trace to cut the memory to the proper size (by checking the observed memory size of the previous step).  This way each step in the geth trace will have the memory size before the step.  Would that solve the issue?

File: bus-mapping/src/evm/memory.rs
Comment (CPerezz) [reply bus-mapping/src/evm/memory.rs:276]: So if you need to take care when there's leftover in the division, there should be extra logic in the fn no?

As otherways you're purely just truncating.

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:33]: I think something like this would fix the memory sizes of the geth exec steps;
```rust
fn fix_geth_trace_memory_size(trace: &mut GethExecTrace) {
    let mut mem_sizes = vec![0; trace.struct_logs.len()];
    let mut call_mem_size_stack = Vec::new();
    for i in 1..trace.struct_logs.len() {
        let step_prev = &trace.struct_logs[i - 1];
        let step = &trace.struct_logs[i];
        mem_sizes[i] = match step.depth as isize - step_prev.depth as isize {
            // Same call context
            0 => step_prev.memory.0.len(),
            // into new call context
            1 => {
                call_mem_size_stack.push(step_prev.memory.0.len());
                0
            }
            // return from call context
            -1 => call_mem_size_stack.pop().expect("call stack is empty"),
            _ => unreachable!(),
        };
    }
    for i in 0..trace.struct_logs.len() {
        trace.struct_logs[i].memory.0.truncate(mem_sizes[i]);
    }
}
```
We could have this transformation in the `GethExecTrace` deserialization step for example.  What do you think?

File: bus-mapping/src/evm/memory.rs
Comment (ed255) [reply bus-mapping/src/evm/memory.rs:276]: I think the memory size will always be multiple of 32.  Internally we represent the memory in bytes, but geth returns it as an array of words (where each word is 32 bytes)

Comment (spartucus) [reply bus-mapping/src/evm/memory.rs:276]: >geth returns it as an array of words (where each word is 32 bytes)

OK, in this case we could use `self.0.len() >> 5`.

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (han0110) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:33]: yeah, using previous step will break when it returns and then use the memory size instantly. Using call stack style to fix this looks great!

Comment (spartucus) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:33]: I tried https://github.com/appliedzkp/zkevm-circuits/commit/432069605b3e7eeec806eda3f313b60edc533e4a, sadly it did not solve the problem. The traced data looks good to me.

The cause is in witness.rs [step_convert](https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/witness.rs#L388-L403):
```rust
execution_state: ExecutionState::from(step), // MSIZE
...
memory_size: match prev {
    None => 0, // MSIZE
    Some(prev_step) => (prev_step.memory_size as u64) / 32,  // STOP
},
```
`MSIZE step` has 0 memory_size, and `STOP` step has proper memory_size that `MSIZE` wanted.

So I add a special case `memory_size` assignment for `MSIZE`. What do you think?

**Comment by spartucus** [comment]
I added a `MSIZE` example in geth-utils,  the core code is:
```Go
asm := gethutil.NewAssembly().MStore(0x40, 0x80).MSize().Stop()
```

And the output is:
```JSON
[
  {
    "pc": 0,
    "op": "PUSH1",
    "gas": 100,
    "gasCost": 3,
    "depth": 1,
    "stack": [],
    "memory": []
  },
  {
    "pc": 2,
    "op": "PUSH1",
    "gas": 97,
    "gasCost": 3,
    "depth": 1,
    "stack": [
      "0x80"
    ],
    "memory": []
  },
  {
    "pc": 4,
    "op": "MSTORE",
    "gas": 94,
    "gasCost": 12,
    "depth": 1,
    "stack": [
      "0x80",
      "0x40"
    ],
    "memory": [
      "0000000000000000000000000000000000000000000000000000000000000000",
      "0000000000000000000000000000000000000000000000000000000000000000",
      "0000000000000000000000000000000000000000000000000000000000000000"
    ]
  },
  {
    "pc": 5,
    "op": "MSIZE",
    "gas": 82,
    "gasCost": 2,
    "depth": 1,
    "stack": [],
    "memory": [
      "0000000000000000000000000000000000000000000000000000000000000000",
      "0000000000000000000000000000000000000000000000000000000000000000",
      "0000000000000000000000000000000000000000000000000000000000000080"
    ]
  },
  {
    "pc": 6,
    "op": "STOP",
    "gas": 80,
    "gasCost": 0,
    "depth": 1,
    "stack": [
      "0x60"
    ],
    "memory": [
      "0000000000000000000000000000000000000000000000000000000000000000",
      "0000000000000000000000000000000000000000000000000000000000000000",
      "0000000000000000000000000000000000000000000000000000000000000080"
    ]
  }
]

```

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (ed255) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:31]: I've taken a look and it seems it didn't work because the external_tracer works directly with `Vec<GethExecStep`, so the Deserialize on `GethExecTrace` is not called.  We can call `fix_geth_trace_memory_size` in the tracer function of the external_tracer.

I've updated the commit I shared before.  Can you try with this? https://github.com/appliedzkp/zkevm-circuits/commit/a86222d727fd15f0a69cc921d3fa787263703ed7
I've also added a test with your example, and now the `MSTORE` has memory size at 0.

**Comment by ed255** [comment]
> I added a `MSIZE` example in geth-utils, the core code is:
> 
> [...]
> 
> So when we `build_block_from_trace_code_at_start`, we get last two object MSIZE and STOP, which they all get a memory_size of 3.

When applying https://github.com/appliedzkp/zkevm-circuits/commit/a86222d727fd15f0a69cc921d3fa787263703ed7 (which fixes the geth trace in rust, both in the Deserialize of GethExecTrace and in the tracer from external_tracer), with the following test:
```rust
        let code = bytecode! {
            #[start]
            PUSH1(0x80)
            PUSH1(0x40)
            MSTORE
            MSIZE
            STOP
        };

        let block =
            mock::BlockData::new_single_tx_trace_code_at_start(&code).unwrap();
        println!("{:#?}", block.geth_trace.struct_logs);
```
We get this is the result (notice that we're using custom Debug impl for the types, so the output is different than original json trace):
```
[
    Step {
        pc: 0x0000,
        op: PUSH1,
        gas: 1000000,
        gas_cost: 3,
        depth: 1,
        error: None,
        stack: [],
        memory: [
        ],
        storage: {},
    },
    Step {
        pc: 0x0002,
        op: PUSH1,
        gas: 999997,
        gas_cost: 3,
        depth: 1,
        error: None,
        stack: [
            0x80,
        ],
        memory: [
        ],
        storage: {},
    },
    Step {
        pc: 0x0004,
        op: MSTORE,
        gas: 999994,
        gas_cost: 12,
        depth: 1,
        error: None,
        stack: [
            0x80,
            0x40,
        ],
        memory: [
        ],
        storage: {},
    },
    Step {
        pc: 0x0005,
        op: MSIZE,
        gas: 999982,
        gas_cost: 2,
        depth: 1,
        error: None,
        stack: [],
        memory: [
                00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 80
        ],
        storage: {},
    },
    Step {
        pc: 0x0006,
        op: STOP,
        gas: 999980,
        gas_cost: 0,
        depth: 1,
        error: None,
        stack: [
            0x60,
        ],
        memory: [
                00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000020  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
                00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 80
        ],
        storage: {},
    },
]
```

Now MSTORE has memory size == 0.  Will this resolve the issue?

File: zkevm-circuits/src/evm_circuit/execution/msize.rs
Comment (spartucus) [reply zkevm-circuits/src/evm_circuit/execution/msize.rs:31]: I changed as https://github.com/appliedzkp/zkevm-circuits/commit/a86222d727fd15f0a69cc921d3fa787263703ed7 and tried yestoday, still same.

**Comment by spartucus** [comment]
No, and the `memory.rs` tests failed too. It looks like something wrong on https://github.com/appliedzkp/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/util/constraint_builder.rs#L313:
```
(
  "State transition constrain of memory_size",
  &self.curr.state.memory_size,
  &self.next.state.memory_size,
  step_state_transition.memory_size,
),
```

**Comment by ed255** [comment]
> And after I change the `step_convert` code in witness.rs to below, both MSIZE and MSTORE/MLOAD in memory.rs tests pass the tests:
> 
> ```rust
> memory_size: step.memory_size as u64 / 32,
> ```
> 
> Is this information is helpful to you?

Yes!  Thea idea is to apply the commit https://github.com/appliedzkp/zkevm-circuits/commit/a86222d727fd15f0a69cc921d3fa787263703ed7 along with that change!  Since now the trace contains the proper memory size at each step, there's no need to use the previous step in the circuit for memory_size.

**Comment by spartucus** [comment]
Ah~ Now I get your point.

**Comment by spartucus** [comment]
@ed255 I think it's ready to review😀

**Review by ed255 — APPROVED** [review]

**Comment by ed255** [comment]
> @ed255 I think it's ready to review :grinning:

LGMT!  @han0110 could you complete the review please?  I'm not yet very familiar with the circuits part so I think it would be good if you can review (and approve if you see fit) the changes.

**Review by han0110 — APPROVED** [review]
LGTM! Tho I still think the `memory_size` could be capped in 4 bytes, but let me unify the number of bytes about memory address and size across different places in further PRs.

**Comment by spartucus** [comment]
Thank you guys for kind help😀

**Pull request merged**
